libvmaf/src/feature/funque_adm.c

```diff
@@ -0,0 +1,319 @@
+					masking_signal[index] = fabsf(pyr_2.bands[k][index]);
+				}
+			}
+		  integral_image_adm_sums(masking_signal, 3, 1, integral_sum, width, height);
+			for (i = 0; i < height; i++)
+			{
+				for (j = 0; j < width; j++)
+				{
+					index = i * width + j;
+					masking_threshold[index] += (integral_sum[index] + masking_signal[index]) / 30;
+				}
+			}
+      }			
+      
+      for (k = 1; k < 4; k++)
+      {
+		  for (i = 0; i < height; i++)
+		  {
+			  for (j = 0; j < width; j++)
+			  {
+				  index = i * width + j;
+				  val = fabsf(pyr_1.bands[k][index]) - masking_threshold[index];
+				  masked_pyr.bands[k][index] = clip(val, 0.0, val);
+			  }
+		  }
+      }
+	  free(masking_signal);
+	  free(masking_threshold);
+	  free(integral_sum);
+}
+
+void dlm_contrast_mask(dwt2buffers pyr_1, dwt2buffers pyr_2, dwt2buffers masked_pyr_1, dwt2buffers masked_pyr_2, size_t width, size_t height)
+{
+  dlm_contrast_mask_one_way(pyr_1, pyr_2, masked_pyr_1, width, height);
+  dlm_contrast_mask_one_way(pyr_2, pyr_1, masked_pyr_2, width, height);
+}
+
+int compute_adm_funque(dwt2buffers ref, dwt2buffers dist, double *adm_score, double *adm_score_num, double *adm_score_den, size_t width, size_t height, funque_dtype border_size)
+{
+  // TODO: assert len(pyr_ref) == len(pyr_dist),'Pyramids must be of equal height.'
+
+  int n_levels = 1;
+  int i, j, k, index;
+  double num_sum = 0, den_sum = 0, num_band = 0, den_band = 0;
+  dwt2buffers dlm_rest, dlm_add, pyr_rest, pyr_add;
+  dlm_rest.bands[0] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  dlm_rest.bands[1] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  dlm_rest.bands[2] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  dlm_rest.bands[3] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  dlm_add.bands[0] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  dlm_add.bands[1] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  dlm_add.bands[2] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  dlm_add.bands[3] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_rest.bands[0] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_rest.bands[1] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_rest.bands[2] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_rest.bands[3] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_add.bands[0] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_add.bands[1] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_add.bands[2] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_add.bands[3] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+
+  dlm_decouple(ref, dist, dlm_rest, dlm_add);
+
+  dlm_contrast_mask(dlm_rest, dlm_add, pyr_rest, pyr_add, width, height);
```

Comment:
A one-way contrast mask is enough because only `pyr_rest` is used after this step, not `pyr_add`. So, you can also skip allocating memory for `pyr_add`.

Reply:
Thanks for pointing out. This redundancy is removed now.

Reply:
Thanks @JeevaRajIttiam!